src/main/java/org/traccar/protocol/FlexApiProtocolDecoder.java

```diff
@@ -56,42 +58,59 @@ protected Object decode(
 
         if (topic.contains("/gnss/")) {
 
-            position.setValid(true);
-
-            if (payload.containsKey("time")) {
-                position.setTime(new Date(payload.getInt("time") * 1000L));
-                position.setLatitude(payload.getJsonNumber("lat").doubleValue());
-                position.setLongitude(payload.getJsonNumber("log").doubleValue());
+            if (payload.getInt("gnss.fix") > 0) {
+                position.setValid(true);
+                if (payload.containsKey("time")) {
+                    position.setTime(new Date(payload.getInt("time") * 1000L));
+                    position.setLatitude(payload.getJsonNumber("lat").doubleValue());
+                    position.setLongitude(payload.getJsonNumber("log").doubleValue());
+                } else {
+                    position.setTime(new Date(payload.getInt("gnss.ts") * 1000L));
+                    position.setLatitude(payload.getJsonNumber("gnss.latitude").doubleValue());
+                    position.setLongitude(payload.getJsonNumber("gnss.longitude").doubleValue());
+                }
+                Optional.ofNullable(payload.getJsonNumber("gnss.altitude"))
+                        .map(JsonNumber::doubleValue).ifPresent(position::setAltitude);
```

Comment:
Personally I think this format is hard to read. If you want to do the check, I would recommend doing it explicitly.

Are you sure all the parameters you check for null are optional?

Reply:
I'll go confirm with firmware developer again.